### PR TITLE
audiounit: add AUConverter resampler (recorder)

### PR DESCRIPTION
**Pull request by nicotyze**
This PR is a proposal of an AUConverter resampler for audiounit module (recorder side).

The approach consists in using a specific AudioUnit for conversion (sub-type AUConverter).
The raw audio samples are captured (at hw sample rate) and then, the conversion is called

I did it in a really empirical way, mixing different "sources" (mainly from [here](http://dinhe.net/~aredridel/projects/audio/audioengine-1.4/ ) and [here](https://github.com/alfredh/baresip/tree/audiounit_resample2))

Tested successfully from OSX 10.13: auloop and calls with many combinations (44100 Hz -> 48000 Hz, 44100 Hz -> 8000 Hz, 48000 Hz -> 44100 Hz ...). 

Not tested from iOS...

### CODE REVIEW

**Comment by alfredh** [comment]
@nicotyze this is really amazing work :)

I will do some testing on OSX and review the code.

**Comment by alfredh** [comment]
there is one problem here, which is the same problem as I had.
when the sample-rate conversion is not an integer number, some samples
are lost.

I added some printf's in the code:

```
inNumberFrames: 512
outNumberFrames: 557
conv: inNumberFrames: 512

inNumberFrames: 512
outNumberFrames: 557
conv: inNumberFrames: 511         <--------- 1 sample lost
```

you can hear this as soft crackling background noise, if you make
a continuous tone in the recorder.

[audiounit.wav.zip](https://github.com/alfredh/baresip/files/2721569/audiounit.wav.zip)

**Comment by nicotyze** [comment]
Hi,
@alfredh, thanks for this nice vocal performance  :-)
Indeed I did not test it deeply enough  in such a way I could have detected this issue ...

Float to integer casting lowers  the value of outNumberFrames and in order to keep a constant conversion ratio (based on in/out sample rate set initially), the converter adapts the number of input samples.
Lowering outNumberFrames with nteger casting

File: modules/audiounit/recorder.c [hunk 1]
```diff
@@ -66,37 +74,79 @@ static OSStatus input_callback(void *inRefCon,
...
+	outNumberFrames = inNumberFrames * st->sampc_ratio;
+	abl_conv.mNumberBuffers = 1;
+	abl_conv.mBuffers[0].mNumberChannels = st->ch;
+	abl_conv.mBuffers[0].mData = NULL;
+	abl_conv.mBuffers[0].mDataByteSize = outNumberFrames * st->sampsz;
+
+	ret = AudioUnitRender(st->au_conv,
+			      ioActionFlags,
+			      inTimeStamp,
+			      0,
```
[the remark this answers is not in the document]
Comment (nicotyze) [reply modules/audiounit/recorder.c]: From this [programming guide](https://developer.apple.com/library/archive/documentation/MusicAudio/Conceptual/AudioUnitProgrammingGuide/TheAudioUnit/TheAudioUnit.html), I understand that a "typical" processing (not I/O) unit has only one bus: 
"In the typical case, the input and output scopes each have one element, namely element (or bus) 0"

I propose to use a bus constant called "defaultBus" in this case...

**Comment by nicotyze** [comment]
Hi,
I have updated the patch.
As mentioned above, it consists in using a specific buffer between the capturer and the converter.
I did it in a very basic way with a kind of double buffer. A circular buffer would probably be more conventional but also heavier in terms of implementation.
This seems to solve the issue, no more noisy sounds from my side.

@alfredh 
I guess this buffering could also work with AudioConverterFillComplexBuffer [approach](https://github.com/alfredh/baresip/blob/audiounit_resample2/modules/audiounit/recorder.c).
I don't know if there is a preferable solution between using AudioConverterFillComplexBuffer and using a specific AUConverter unit as I did...

**Comment by alfredh** [comment]
great work, just tested it and it works better :)

general comments:

1. the buffer will most likely also be used by the playback, so perhaps
  we should move the buffer type and code to a shared file,
  such as audiounit.c or buffer.c

2. when testing auloop this is what I get with 48000Hz:

```
read:  1 sampc
read:  557 sampc
read:  557 sampc
read:  557 sampc
read:  1 sampc
read:  557 sampc
read:  557 sampc
read:  557 sampc
read:  557 sampc
read:  1 sampc
```

perhaps it is more efficient to fill up the buffer until `ptime` milliseconds 
and then pass the buffer to the application.

for example 8000 Hz at 20ms gives 160 samples per block.

the wanted sample count can be calculated like this:

```
sampc = prm->srate * prm->ch * prm->ptime / 1000
```

File: modules/audiounit/recorder.c [hunk 2]
```diff
@@ -33,16 +44,67 @@ static void ausrc_destructor(void *arg)
...
+	if (st->conv_buf.nb_frames < nb_frames) {
+		return kAudioUnitErr_TooManyFramesToProcess;
+	}
+
+	*data = st->conv_buf.mem[mem_idx];
+
+	delta = st->conv_buf.nb_frames - nb_frames;
+
+	uint8_t *src = (uint8_t *)st->conv_buf.mem[mem_idx] +
+		       nb_frames * st->sampsz;
```
Comment (alfredh) [inline modules/audiounit/recorder.c]: `src` should be declared at the beginning of the function.

**Comment by nicotyze** [comment]
Hi @alfredh,
Thanks for the review and comments.
There is a bug in the patch, related to data size calculation (I missed the channels number...).
I'm fixing it and preparing a new commit.

**Comment by alfredh** [comment]
@nicotyze great, thank you

**Comment by nicotyze** [comment]
@alfredh 
I've just pushed a new commit following your recommendations.
About the second point I've just modified the while loop in order to call AudioUnitRender with inNumberFrames as input,  keeping it  as close as possible to the initial implementation.
My understanding is that waiting ptime milliseconds before passing the buffer to the application would require an other buffer between the converter and the application.

**Comment by alfredh** [comment]
@nicotyze thanks, I merged your patch to master.

Thanks a bunch for this work! This is greatly appreciated :)

**Comment by Koshub** [comment]
@nicotyze @alfredh Could you, please, explain why the converter is used, since as I know the the output of the input accepts the application/client format to covert to and can convert to that format from hardware format including the sample rate?

**Comment by alfredh** [comment]
it was added because the hardware
had to be opened with a specific
sample rate, different from codec

baresip resampler cannot
handle eg 8000 -> 44100 hz

**Comment by Koshub** [comment]
@alfredh Thanks for the reply. So let's assume the input hardware is at 8000 but the codec needs `prm->srate` (say 48000), so we need to resample the input stream to `prm->srate` and write resampled data to the `rh` buffer. Do I understand you correctly? 
If so, according to [this (maybe outdated) doc,](https://developer.apple.com/library/archive/documentation/MusicAudio/Conceptual/AudioUnitHostingGuide_iOS/ConstructingAudioUnitApps/ConstructingAudioUnitApps.html#//apple_ref/doc/uid/TP40009492-CH16-SW2) the IO unit has a converter on both input and output sides under the hood.

> As you can see in the figure, the audio input hardware imposes its stream format on the outward-facing side of the Remote I/O unit’s input element. You, in turn, specify the format that you want to use on the inward-facing side of this element. The audio unit performs format conversion as needed. To avoid unnecessary sample rate conversion, be sure to use the audio hardware sample rate when defining your stream format.

It does not directly explain that the resampling is working but it says `To avoid unnecessary sample rate conversion`. So I think it can resample without separate resampling.

Anyway, it currently works ok with a converter. I will check without it if it makes sense. What do you think?

**Comment by alfredh** [comment]
last time i tested it did
not work..

this apple api is quite difficult
to use correctly

did you test both osx and ios?

**Comment by Koshub** [comment]
I mean since the current implementation is working, I am not sure if it requires to be changed. And yes, it is quite difficult. I did not test yet. Currently, I am just trying to find out the ways to improve this module according to the, let's say, modern possibilities. Once I test different ways I will let you know what I found.

**Pull request merged**